dtos/event.go

```diff
@@ -96,3 +100,101 @@ func (e *Event) ToXML() (string, error) {
 
 	return string(eventXml), nil
 }
+
+func (e Event) MarshalJSON() ([]byte, error) {
+	return e.marshal(json.Marshal)
+}
+
+func (e Event) MarshalCBOR() ([]byte, error) {
+	return e.marshal(cbor.Marshal)
+}
+
+func (e Event) marshal(marshal func(any) ([]byte, error)) ([]byte, error) {
+	var aux struct {
+		dtoCommon.Versionable `json:",inline"`
+		Id                    string        `json:"id" validate:"required,uuid"`
+		DeviceName            string        `json:"deviceName"`
+		ProfileName           string        `json:"profileName"`
+		SourceName            string        `json:"sourceName"`
+		Origin                int64         `json:"origin"`
+		Readings              []BaseReading `json:"readings"`
+		Tags                  Tags          `json:"tags,omitempty"`
+	}
+	aux.Versionable = e.Versionable
+	aux.Id = e.Id
+	aux.DeviceName = e.DeviceName
+	aux.ProfileName = e.ProfileName
+	aux.SourceName = e.SourceName
+	aux.Origin = e.Origin
+	aux.Tags = e.Tags
+	if len(e.Readings) > 0 {
+		aux.Readings = make([]BaseReading, len(e.Readings))
+	}
+
+	if os.Getenv(common.EnvOptimizeEventPayload) == common.ValueTrue {
+		for i, reading := range e.Readings {
+			reading.Id = ""
+			reading.DeviceName = ""
+			reading.ProfileName = ""
+			if e.Origin == reading.Origin {
+				reading.Origin = 0
```

Comment:
Just to confirm, will a zero origin be stripped out the reading when omitempty is specified?

I think the goal of ADR is to remove origin field out of reading when its value is identical to the event-level origin.